### PR TITLE
Workaround for Debug, since omptarget requires -03 to build

### PR DESCRIPTION
**Pull request by PDoakORNL**
## Proposed changes

I've added a build workaround required by the current functional clang 15 offload compiler which cannot build offload target kernels with -O0. This breaks out the omptarget files in ParticleSet and QMCWavefunctions as separate targets. If you actually use a debugger and have experienced using it on -O3 code you'll realize there is no way we want all of ParticleSet and QMCWavefunctions compiled with -O3 in debug.

This also required refactoring BsplineFunctor.h which contained omp target and was ending up in many compilation units.  I also couldn't stand the extremely sloppy template parameter and typedef use in the class and cleaned that up.  While it looked as if T could be any scalar it in fact could only be the QMCTraits::RealType.  There were a few value_type typedefs but they were set at the top of the header using value_type = real_type. 

## What type(s) of changes does this code introduce?

- Bugfix
- Code style update (formatting, renaming)
- Refactoring (no functional changes, no api changes)
- Build related changes
- Other (please describe):

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?

Leconte

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes This PR is up to date with current the current state of 'develop'
- Yes Code added or changed in the PR has been clang-formatted
- Yes This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Yes Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by ye-luo** [comment]
Test this please

**Comment by PDoakORNL** [comment]
More tech debt with sloppy override's and typedefs to pay off in BsplineFunctor stay tuned

File: src/Particle/CMakeLists.txt [hunk 1]
```diff
@@ -53,6 +51,34 @@ target_include_directories(qmcparticle PUBLIC "${CMAKE_CURRENT_SOURCE_DIR}")
 target_link_libraries(qmcparticle PRIVATE platform_cpu_LA)
 target_link_libraries(qmcparticle PUBLIC qmcnumerics qmcutil platform_runtime)
 
+# Workaround for broken OMP Target debug in clang as of 5d, main is more broken so
+# perhaps this problem is fixed.
```
Comment (markdewing) [inline src/Particle/CMakeLists.txt:55]: "As of 5d"? (supposed to be "15"? but that is main, right?)
"main is more broken, so perhaps this problem is fixed" ?

I'm not sure what this comment is trying to say.

Comment (PDoakORNL) [reply src/Particle/CMakeLists.txt:55]: I'm referring to the last known not broken for compiling offload target (although broken for some tooling) clang 15 commit from https://github.com/ye-luo/llvm_recipes/blob/master/olcf_summit/llvm_20220317/upstream-script.sh

File: src/Particle/CMakeLists.txt [hunk 2]
```diff
@@ -53,6 +51,34 @@ target_include_directories(qmcparticle PUBLIC "${CMAKE_CURRENT_SOURCE_DIR}")
...
 target_link_libraries(qmcparticle PUBLIC qmcnumerics qmcutil platform_runtime)
 
+# Workaround for broken OMP Target debug in clang as of 5d2ce7663b10c107328a4ae0c678165209e64619
+# but clang 15 main is completely broken for our omptarget code,
+#
+# In the meantime we need to make a new target for just the omp target tainted files
+# so we don't have broken debug for all of ParticleSet.
+
+# this can be put back into qmcparticle when omp target isn't broken.
+set(PARTICLEOMPTARGET
```
Comment (ye-luo) [inline src/Particle/CMakeLists.txt:61]: Could you rename this to `PARTICLE_OMPTARGET_SRCS` to help reading.

File: src/Particle/CMakeLists.txt [hunk 3]
```diff
@@ -53,6 +51,34 @@ target_include_directories(qmcparticle PUBLIC "${CMAKE_CURRENT_SOURCE_DIR}")
...
+else(USE_OBJECT_TARGET)
+  add_library(qmcparticle_omptarget ${PARTICLEOMPTARGET})
+endif(USE_OBJECT_TARGET)
+
+target_include_directories(qmcparticle_omptarget PUBLIC "${CMAKE_CURRENT_SOURCE_DIR}")
+target_link_libraries(qmcparticle_omptarget PUBLIC qmcutil containers)
+
+if(ENABLE_OFFLOAD AND ${COMPILER} MATCHES "Clang")
+    target_compile_options(qmcparticle_omptarget PRIVATE $<$<CONFIG:DEBUG>:-O3>)
+endif()
```
Comment (ye-luo) [inline src/Particle/CMakeLists.txt:76]: Could you introduce a CMake option `ENABLE_OFFLOAD_CLANG_DEBUG_O3` at top level CMakeLists.txt and repetitive comments next to this option.

File: src/QMCWaveFunctions/CMakeLists.txt [hunk 4]
```diff
@@ -164,6 +166,32 @@ target_include_directories(qmcwfs PUBLIC "${CMAKE_CURRENT_SOURCE_DIR}")
 target_link_libraries(qmcwfs PUBLIC qmcutil qmcparticle platform_runtime)
 target_link_libraries(qmcwfs PRIVATE einspline platform_LA Math::FFTW3)
 
+# Workaround for broken OMP Target debug in clang as of 5d2ce7663b10c107328a4ae0c678165209e64619
+# but clang 15 main is completely broken for our omptarget code,
+#
+# In the meantime we need to make a new target for just the omp target tainted files
+# so we don't have broken debug for all of QMCWavefunctions
```
Comment (ye-luo) [inline src/QMCWaveFunctions/CMakeLists.txt:173]: Remove `workaround... ` which will be at the cmake option.
Here say `# Create qmcwfs_omptarget with source files using OpenMP target offload`
Better with neutral words.

Comment (PDoakORNL) [reply src/QMCWaveFunctions/CMakeLists.txt:173]: I will add neutrality.

**Review by ye-luo — APPROVED** [review]
LGTM

**Comment by ye-luo** [comment]
Test this please

**Pull request merged**